jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py

```diff
@@ -159,47 +203,153 @@ def _on_spectral_unit_changed(self, *args):
+        # skip if arg is not a dictionary
+        if not isinstance(msg, dict):
+            raise TypeError("Expected msg to be a dictionary")
+        # skip if 'name' key is missing
+        name = msg.get('name')
+        if not name:
+            raise KeyError("Expected 'name' key in msg")
+
+        # determine if flux or surface brightness unit was changed by user
+        if name == 'flux_unit_selected':
+            # when the configuration is Specviz, translation is not currently supported.
+            # If in Cubeviz, all spectra pass through Spectral Extraction plugin and will
+            # have a scale factor assigned in the metadata, enabling translation.
+            if data_collection_unit == 'Surface Brightness':
+                raise ValueError(
+                    f"Unit translation between Flux and Surface Brightness "
+                    f"is not supported in {self.app.config}."
+                )
+            flux_or_sb = self.flux_unit.selected
+            # update flux or surface brightness dropdown if necessary
+            if check_if_unit_is_per_solid_angle(current_y):
+                self._translate('Flux')
+                self.flux_or_sb.selected = 'Flux'
+            untranslatable_units = self._untranslatable_units
+            # disable translator if flux unit is untranslatable,
+            # still can convert flux units, this just disables flux
+            # to surface brightnes translation for units in list.
+            if flux_or_sb in untranslatable_units:
+                self.can_translate = False
+            else:
+                self.can_translate = True
+
+        elif name == 'sb_unit_selected':
+            if data_collection_unit == 'Flux':
+                # when the configuration is Specviz, translation is not currently supported.
+                # If in Cubeviz, all spectra pass through Spectral Xxtraction plugin and will
+                # have a scale factor assigned in the metadata, enabling translation.
+                raise ValueError(
+                    f"Unit translation between Flux and Surface Brightness "
+                    f"is not supported in {self.app.config}."
+                )
+            flux_or_sb = self.sb_unit.selected
+            self.can_translate = True
+            # update flux or surface brightness dropdown if necessary
+            if not check_if_unit_is_per_solid_angle(current_y):
+                self._translate('Surface Brightness')
+                self.flux_or_sb.selected = 'Surface Brightness'
+        elif name == 'flux_or_sb_selected':
+            self._translate(self.flux_or_sb_selected)
+            return
+        else:
+            return
+
+        yunit = _valid_glue_display_unit(flux_or_sb, self.spectrum_viewer, 'y')
+        # may need to be updated if translations in other configs going to be supported
+        if not self.flux_unit.choices and self.app.config == 'cubeviz':
```

Comment:
I don't see why it can't happen earlier, let me test it out earlier in the `_on_flux_unit_changed` and I'll make the change bearing no behavioral change occurs!

Reply:
I do get an AttributeError: 'UnitConversion' object has no attribute 'flux_unit' if I move it further up. I can spend some time seeing if/what I can change the conditional to to have the same behavior.

Reply:
you can add a `if not hasattr(self, 'flux_unit'): return` to the top of the method if you'd like.  The observe on `flux_unit_selected` can get triggered when the traitlet is created but before the plugin's init has created the `flux_unit` object itself, and there's no need to run all this logic at that point.

Reply:
I thought I had tried the not hasattr before, but irregardless that seemed to do the trick and it is now at the top of the function. Just running through tests locally now to make sure it.

Reply:
The `flux_unit` can be created, but the `self.flux_unit.choices` isn't populated yet causing one of the AttributeErrors, but the actually choices don't change once populated.